### PR TITLE
EWS for Office 365: PAPP-31110 - Bug fix for encrypting token longer than 4096 characters

### PR DESCRIPTION
**Pull request by ishans-crest**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [x] The readme.html has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [x] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [x] Verify all checks are passing.
- [ ] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [ ] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe): 

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers), 
  please document them in the `readme.html`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `readme.html` all methods (eg, OAuth) used to authenticate 
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `readme.html`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through 
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

  

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

**Review by splunk-jessica — APPROVED** [review]

File: ewsonprem_connector.py [hunk 1]
```diff
@@ -508,6 +506,23 @@ def _get_domain(self, username, client_req_id):
...
+        if self.get_action_identifier() != phantom.ACTION_ID_TEST_ASSET_CONNECTIVITY:
+            resp_json = None
+            if self._state.get("oauth_token", {}):
+                resp_json = self._state.get("oauth_token", {})
+            if self._state.get("oauth_client_token", {}):
+                resp_json = self._state.get("oauth_client_token", {})
+            if resp_json:
+                self._session.auth = OAuth2TokenAuth(resp_json['access_token'], resp_json['token_type'])
+        # self._is_token_test_connectivity variable set for generating the tokens only for once while having
+        # the multiple calls.
```
Comment (splunk-jessica) [inline ewsonprem_connector.py:520]: not sure what this comment means. If it's not necessary, can we remove it?

File: ewsonprem_connector.py [hunk 2]
```diff
@@ -749,30 +776,44 @@ def _decrypt_client_token(self, state):
...
             self.debug_print("Resetting the state file with the default format")
             self._state = {"app_version": self.get_app_json().get("app_version")}
             if self.auth_type == AUTH_TYPE_AZURE_INTERACTIVE:
                 return self.set_status(phantom.APP_ERROR, EWSONPREM_STATE_FILE_CORRUPT_ERROR)
+        if self.auth_type == AUTH_TYPE_CLIENT_CRED:
+            self._state = self._decrypt_client_token(self._state)
+        else:
+            self._state, message = self.rsh._decrypt_state(self._state)
+            if message:
+                # self.save_progress(message)
```
Comment (splunk-jessica) [inline ewsonprem_connector.py:815]: remove comment

File: ewsonprem_connector.py [hunk 3]
```diff
@@ -943,12 +964,49 @@ def _get_http_error_details(self, r):
 
         return ""
 
+    def all_authentication_methods(self, config):
```
Comment (splunk-jessica) [inline ewsonprem_connector.py:967]: Can this function name be changed to something more descriptive? Maybe `set_authentication_method` would be better?

File: ewsonprem_connector.py [hunk 4]
```diff
@@ -943,12 +964,49 @@ def _get_http_error_details(self, r):
...
+        "Method for using all type of authentication"
```
Comment (splunk-jessica) [inline ewsonprem_connector.py:968]:
```suggestion
        "Method for setting authentication"
```

File: ewsonprem_connector.py (hunk 2 above)
Comment (ishans-crest) [reply ewsonprem_connector.py:815]: Removed

File: ewsonprem_connector.py (hunk 3 above)
Comment (ishans-crest) [reply ewsonprem_connector.py:967]: Changed the definition name to set_authentication_method

File: ewsonprem_connector.py (hunk 1 above)
Comment (ishans-crest) [reply ewsonprem_connector.py:520]: Removed it

**Review by ricardo-phantom — APPROVED** [review]

**Comment by ishans-crest** [comment]
Prodsec failures are discussed in slack hence merging the PR

**Review by alexa-phantom — APPROVED** [review]

**Pull request merged**